src/shims/unix/sync.rs

```diff
@@ -90,79 +90,120 @@ fn mutex_id_offset<'tcx>(ecx: &MiriInterpCx<'tcx>) -> InterpResult<'tcx, u64> {
-        let static_initializer = ecx.eval_path(&["libc", "PTHREAD_MUTEX_INITIALIZER"]);
-        let kind_field = static_initializer
-            .offset(Size::from_bytes(mutex_kind_offset(ecx)), ecx.machine.layouts.i32, ecx)
-            .unwrap();
-        let kind = ecx.read_scalar(&kind_field).unwrap().to_i32().unwrap();
-        assert_eq!(
-            kind,
-            ecx.eval_libc_i32("PTHREAD_MUTEX_DEFAULT"),
-            "PTHREAD_MUTEX_INITIALIZER is incompatible with our pthread_mutex layout: kind is not PTHREAD_MUTEX_DEFAULT"
-        );
-    }
-
-    offset
+/// Eagerly create and initialize a new mutex.
+fn mutex_create<'tcx>(
+    ecx: &mut MiriInterpCx<'tcx>,
+    mutex_op: &OpTy<'tcx>,
+    kind: i32,
+) -> InterpResult<'tcx> {
+    let address = ecx.read_pointer(mutex_op)?.addr().bytes();
+    let kind = MutexKind::new(kind);
+    let data = Some(AdditionalMutexData { address, kind });
+    ecx.mutex_create(mutex_op, ecx.libc_ty_layout("pthread_mutex_t"), mutex_id_offset(ecx)?, data)?;
+    Ok(())
 }
 
+/// Returns the `MutexId` of the mutex stored at `mutex_op`.
+///
+/// `mutex_get_id` will also check if the mutex has been moved since its first use and
+/// return an error if it has.
 fn mutex_get_id<'tcx>(
     ecx: &mut MiriInterpCx<'tcx>,
     mutex_op: &OpTy<'tcx>,
 ) -> InterpResult<'tcx, MutexId> {
-    ecx.mutex_get_or_create_id(
+    // FIXME: might be worth changing mutex_get_or_create_id to take the mplace
+    // rather than the `OpTy`.
+    let id = ecx.mutex_get_or_create_id(
         mutex_op,
         ecx.libc_ty_layout("pthread_mutex_t"),
         mutex_id_offset(ecx)?,
-    )
+        |ecx| {
+            let address = ecx.read_pointer(mutex_op)?.addr().bytes();
+
+            let mutex = ecx.deref_pointer(mutex_op)?.ptr();
+            let contents = get_contents(ecx, mutex)?;
+            let kind = kind_from_contents(ecx, contents)?;
+
+            Ok(Some(AdditionalMutexData { kind, address }))
+        },
+    )?;
+    mutex_check_moved(ecx, mutex_op, id)?;
+    Ok(id)
 }
 
-fn mutex_reset_id<'tcx>(
-    ecx: &mut MiriInterpCx<'tcx>,
-    mutex_op: &OpTy<'tcx>,
-) -> InterpResult<'tcx, ()> {
-    ecx.deref_pointer_and_write(
-        mutex_op,
-        mutex_id_offset(ecx)?,
-        Scalar::from_u32(0),
-        ecx.libc_ty_layout("pthread_mutex_t"),
-        ecx.machine.layouts.u32,
-    )
+/// Returns the kind of a mutex, based on its contents.
+/// The contents must not contain the initial "header" where the mutex id is stored.
+fn kind_from_contents<'tcx>(
+    ecx: &MiriInterpCx<'tcx>,
+    contents: &[u8],
+) -> InterpResult<'tcx, MutexKind> {
+    let on_linux = &*ecx.tcx.sess.target.os == "linux";
+
+    let default_initializer =
+        ecx.eval_path(&["libc", "PTHREAD_MUTEX_INITIALIZER"]).assert_mem_place().ptr();
+    if contents == get_contents(ecx, default_initializer)? {
+        return Ok(MutexKind::new(ecx.eval_libc_i32("PTHREAD_MUTEX_DEFAULT")));
+    }
```

Comment:
This still makes the assumption that the platform-specific field that encodes the mutex kind in the static initializer is after the "header". So it's not enough to be sure we are matching platform behavior, and it's not clean enough IMO to keep it as a half-measure.

I also just realized that this function is also called *after* the new ID has been written, so a full comparison with `pthread_mutex_t` would anyway not work.

So I think the best option is to have a `match &*ecx.tcx.sess.target.os` inside this function and then determine for each platform what we have to read to find the initial mutex kind -- basically exactly what `mutex_get_kind` did before this PR, but only doing it here when we "parse" a static initializer, rather than doing it each time the mutex is used. On Linux this would read 4 bytes at offset `if ecx.pointer_size().bytes() == 8 { 16 } else { 12 }`. For the other supported targets we'll have to check whether they have any static initializers other than PTHREAD_MUTEX_DEFAULT. And if any platform adds such initializers in the future, Miri will just ignore them until we realize that... maybe one day we can do the comparison with `PTHREAD_MUTEX_DEFAULT` but it will require reworking `get_or_create_id` a bit more so that `create_obj` is called before the new ID is written.

Reply:
I checked and none of the other targets has any other static initializers.

Reply:
done. I am not sure I interpreted all of your comment correctly, so let me know.